### PR TITLE
gui: Use case-insensive and backslash-agnostic versions filter (fixes #7973)

### LINKED ISSUE #7973: Restore Versions file filter should have an option to disable case sensitivity
Currently, the filter is case **sensitive**.

image

I'd argue that there should be an option to make it case insensitive (perhaps in a form of a checkbox). Alternatively, the filter could automatically adapt to the operating system at hand (e.g. as the ignore patterns do, being case insensitive on Windows).

### PR DESCRIPTION
**Pull request by tomasz1986**
gui: Use case-insensive versions filter and allow backslashes as path separators (fixes #7973)

Currently, the versions filter is case-sensitive regardless of the
underlying OS. With this change, the filter becomes case-insensitive
everywhere, which is more user-friendly and makes it easier to search
for files whose exact case the user may not remember.

In addition, allow backslashes to be used as filter path separators.

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>

### CODE REVIEW

**Comment by calmh** [comment]
This filter is just used for what we see when typing something, right, not used for actual file system interaction? In that case, would it make sense to always have it case insensitive? I think that would be more human centric, I don't think it'd be annoyed by a view filter being case insensitive even on Linux.

**Comment by tomasz1986** [comment]
> This filter is just used for what we see when typing something, right, not used for actual file system interaction? In that case, would it make sense to always have it case insensitive? I think that would be more human centric, I don't think it'd be annoyed by a view filter being case insensitive even on Linux.

Yeah, this is only used for filtering versioned files as they are displayed in the GUI. I can make it always case-insensitive, of course, if that's more desirable. This way, even if there are multiple files that only differ in case, the filter will just always display all of them when matched, regardless of the case.

**Comment by acolomb** [comment]
Full-time Linux user here.  Yes, please make it *always case-insensitive*.  Searching for an entry in a view by entering filter criteria would be less useful and surprising if I had to match case exactly.  File managers do the same, as users tend to only enter some characters without thinking about capitalization.

**Comment by tomasz1986** [comment]
I've changed the filter to always be case-insesitive!

File: gui/default/syncthing/core/syncthingController.js [hunk 1]
```diff
@@ -2778,9 +2778,21 @@ angular.module('syncthing.core')
...
-                    return false;
+
+                if ($scope.restoreVersions.filters.text) {
+                    // Use case-insensitive filter.
+                    var filterText = $scope.restoreVersions.filters.text.toLowerCase();
+                    var versionPath = node.key.toLowerCase();
+                    // Allow backslashes on Windows as filter path separators.
+                    if ($scope.version.os == 'windows') {
+                        filterText = filterText.replace(/\//g, '\\');
+                        versionPath = versionPath.replace(/\//g, '\\');
```
Comment (acolomb) [inline gui/default/syncthing/core/syncthingController.js:2789]: If I understand this correctly, you're replacing every slash with a backslash on Windows?  The regex is kinda hard to read because of all the slashes involved for different purposes.

Why not go the other way round?  Doesn't Syncthing use forward-slashes for relative paths within a folder?  Converting backslashes to slashes would make more sense to me, as it reduces differences between OSes.

Comment (tomasz1986) [reply gui/default/syncthing/core/syncthingController.js:2789]: Wouldn't this interfere with using backslashes for escape on other OSs, e.g. `my\ file`? As far as I'm aware, you can also have backslashes legitimately as a part of the filename too (but not in Windows).

Comment (imsodin) [reply gui/default/syncthing/core/syncthingController.js:2789]: Afaik we use native path separators here, so `/` everywhere except windows where it's `\\`. However that would make this change a noop on the `versionPath` side of things, so I guess you checked and it's using forward slashes here @tomasz1986 ? I had a quick look at the code and didn't see anything that suggest that we use forward slashes everywhere, but I might have easily missed the place where a translation happens.

Comment (tomasz1986) [reply gui/default/syncthing/core/syncthingController.js:2789]: Yeah, the tree uses forward slashes which I think come from here:

https://github.com/syncthing/syncthing/blob/f42f041f5329f08117624d79fe25ed5e695f5e85/gui/default/syncthing/app.js#L176-L214

Comment (tomasz1986) [reply gui/default/syncthing/core/syncthingController.js]: @acolomb I've changed the conversion as you asked and improved the comment to make it clear what is going on!

**Review by acolomb — APPROVED** [review]
I haven't tested this myself, as I have no appropriate setup involving versioning and Windows hosts.  But I trust your judgment there, maybe give some examples how it works in the PR description?

**Comment by acolomb** [comment]
Sorry, scratch my last comment.  I overlooked that you are still activating this logic only for Windows.  That is, if the GUI runs in a browser on Windows, right?  But the paths may be from a different OS if accessing a remote GUI.  I think the OS check should just be skipped.

If one searches for something containing a slash, it shouldn't matter whether it's a forward- or backslash.  The quoting you mentioned happens in context of UNIX shells (e.g. to distinguish an argument containing a space from a space separating two arguments), it's not inherent to the file systems.  So don't worry about that.

File: gui/default/syncthing/core/syncthingController.js (hunk 1 above)
Comment (imsodin) [reply gui/default/syncthing/core/syncthingController.js]: > Yeah, the tree uses forward slashes which I think come from here:

Then there's no need/point in changing `versionPath`, right?

Comment (acolomb) [reply gui/default/syncthing/core/syncthingController.js]: Except if someone expects to paste a (partial) Windows path into the search box and have it match hierarchical paths with their forward slashes.  Thus I'd keep it.

**Comment by tomasz1986** [comment]
> Sorry, scratch my last comment. I overlooked that you are still activating this logic only for Windows. That is, if the GUI runs in a browser on Windows, right? But the paths may be from a different OS if accessing a remote GUI. I think the OS check should just be skipped.

The OS check is actually about Syncthing itself, meaning that it's applied only when running the Windows binary of Syncthing.

> If one searches for something containing a slash, it shouldn't matter whether it's a forward- or backslash. The quoting you mentioned happens in context of UNIX shells (e.g. to distinguish an argument containing a space from a space separating two arguments), it's not inherent to the file systems. So don't worry about that.

But on Linux and macOS, you can have literal `\` in filenames, right? Of course, it's very unlikely, but still converting them all to `/` will change the meaning… On Windows this doesn't matter, because backslash is the path separator and forward slash is forbidden in filenames anyway.

**Comment by acolomb** [comment]
On Windows, neither slash variant is admitted within file names.  So either should be treated as a path separator (as the OS does itself).

On *NIX etc., either variant **can** be part of a file name, *if* the underlying file system supports it.  But it's still very hard and cumbersome to create such path component names, because the usual GUI tools don't offer it.  I'd say it's a really unlikely situation when somebody tries to shoot their own foot in this way.

So what would happen if we do match a *file name containing* '/' (!) if '\' is searched for?  Worst case, we get a few results too many, that are already named miserably.  And matching a *file name containing* '\' when searching '/'?  Again, a few more results of files which already have poor names.  As for *actual path separators*, I could imagine someone writing down a path in a document or an email using backslashes (because Windows) and me pasting it into the search field.  I'd be surprised if it didn't find anything.

In short, I don't think we need to be exact here.  Think of forward- and backslash as upper and lowercase variants of each other, then it makes total sense.

**Comment by tomasz1986** [comment]
I think I've implemented all the requested changes!

File: gui/default/syncthing/core/syncthingController.js [hunk 2]
```diff
@@ -2778,9 +2778,17 @@ angular.module('syncthing.core')
...
             $scope.restoreVersions.tree.filterNodes(function (node) {
                 if (node.folder) return false;
-                if ($scope.restoreVersions.filters.text && node.key.indexOf($scope.restoreVersions.filters.text) < 0) {
-                    return false;
+
+                if ($scope.restoreVersions.filters.text) {
+                    // Use case-insensitive filter and convert backslashes to
+                    // forward slashes to allow using them as path separators.
+                    var filterText = $scope.restoreVersions.filters.text.toLowerCase().replace(/\\/g, '/');
+                    var versionPath = node.key.toLowerCase().replace(/\\/g, '/');
```
Comment (imsodin) [inline gui/default/syncthing/core/syncthingController.js:2786]: This earlier comment still applies I think (not very important nit, but still :) ):

> > Yeah, the tree uses forward slashes which I think come from here:
> 
> Then there's no need/point in changing `versionPath`, right?

As path separators are already slashes everywhere, there's no need for `replace` on `versionPath`.

Comment (tomasz1986) [reply gui/default/syncthing/core/syncthingController.js:2786]: I think this is still required, as @acolomb explained in details above in https://github.com/syncthing/syncthing/pull/8995#issuecomment-1650642030, because otherwise there will be no way to match a filename that has an actual backslash in its name (for whatever reason).

Comment (imsodin) [reply gui/default/syncthing/core/syncthingController.js:2786]: I don't undrestand how that comment relates to the (internal) paths created from the tree - seems to refer to the user input exclusively. However I might mess things up, at worst it's pointless but even to my (potentially flawed/partial) understanding the `replace` shouldn't hurt.

Comment (tomasz1986) [reply gui/default/syncthing/core/syncthingController.js:2786]: It's not about path separators. On Linux, there is a possibility of having a file named literally `my\file` (with the backslash character in the name, not a path separator). If there is a force replacement of all backslashes to forward slashes in the filter (regardless of the OS) but no replacement in the version path, then it will be impossible to match `my\file` at all.

**Review by acolomb — APPROVED** [review]

**Comment by acolomb** [comment]
@imsodin Okay to merge?

I rethought your comments and I guess it's correct that replacement on the tree's `versionPath` is not necessary *when thinking about path separators*.  But in the general sense, I'd say path separators are just one field where forward and backward slashes are involved and sometimes ambiguous.  Although not possible everywhere and being a footgun where it *is* possible, either slash variant *within* a filename needs special consideration.  So in the same spirit as we decide case doesn't matter, I guess it is more convenient to declare that the direction of slashes doesn't matter either, wherever they are used.
